text_extensions_for_pandas/array/tensor.py

```diff
@@ -0,0 +1,223 @@
+        return TensorArray([a._tensor for a in to_concat])
+
+    def isna(self) -> np.array:
+        """
+        See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        # TODO any or all values in row nan?
+        return np.any(np.isnan(self._tensor), axis=1)
+
+    def copy(self) -> "TensorArray":
+        """
+        See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        ret = TensorArray(
+            self._tensor,
+        )
+        # TODO: Copy cached properties too
+        return ret
+
+    def take(
+        self, indices: Sequence[int], allow_fill: bool = False,
+        fill_value: Any = None
+    ) -> "TensorArray":
+        """
+        See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        # TODO if allow_fill:
+        values = self._tensor.take(indices, axis=0)
+        return TensorArray(values)
+
+    @property
+    def dtype(self) -> pd.api.extensions.ExtensionDtype:
+        """
+        See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        return TensorType()
+
+    def to_numpy(self, dtype=None, copy=False, na_value=pd.api.extensions.no_default):
+        """
+        See docstring in `ExtensionArray` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        # TODO options
+        return self._tensor
+
+    def __len__(self) -> int:
+        return len(self._tensor)
+
+    def __eq__(self, other):
+        return TensorArray(self._tensor == other._tensor)
+
+    def __lt__(self, other):
+        """
+        Pandas-style array/series comparison function.
+
+        :param other: Second operand of a Pandas "<" comparison with the series
+        that wraps this TokenSpanArray.
+
+        :return: Returns a boolean mask indicating which rows are less than
+         `other`. span1 < span2 if span1.end <= span2.begin.
+        """
+        return TensorArray(self._tensor < other._tensor)
+
+    def __gt__(self, other):
+        return TensorArray(self._tensor > other._tensor)
+
+    def __le__(self, other):
+        return TensorArray(self._tensor <= other._tensor)
+
+    def __ge__(self, other):
+        return TensorArray(self._tensor >= other._tensor)
+
+    def __getitem__(self, item) -> "TensorArray":
+        """
+        See docstring in `Extension   Array` class in `pandas/core/arrays/base.py`
+        for information about this method.
+        """
+        # TODO pandas converts series with np.asarray, then applied a function e.g. map_infer(array, is_float) to format strings etc.
+        # If single element return as np.ndarray, if slice return TensorArray of slice
+        if isinstance(item, int):
+            return self._tensor[item]
```

Comment:
This allows a `TensorArray` to be used in `np.asarray` and result in a `ndarray` with the correct dtype. It's needed because I ran into problems otherwise when used as a column inside a DataFrame.

Reply:
Yes, `__getitem__` needs to return the "scalar" type when the index is an `int` and the "array" type when the index is a slice/list/iterable. The API docs don't make this point terribly clear.

Reply:
Ah, ok good then!